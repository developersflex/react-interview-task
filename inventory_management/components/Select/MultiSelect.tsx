import {
  circleDynamicBgStyles,
  getCategoryClassName,
  hoverDynamicBgStyles,
} from "@/utils/functions";
import { useField, useFormikContext } from "formik";
import React, { HTMLAttributes, useEffect, useState } from "react";
import { Icons } from "../Icons";

interface Option {
  label: string;
  value: string;
}

interface MultiSelectProps extends HTMLAttributes<HTMLDivElement> {
  options: Option[];
  label: string;
  placeholder: string;
  name: string;
  className?: string;
}

const MultiSelect: React.FC<MultiSelectProps> = ({
  options,
  label,
  placeholder,
  name,
  className,
}) => {
  const [selectedOptions, setSelectedOptions] = useState<Option[]>([]);
  const [isOpen, setIsOpen] = useState(false);
  const [field, meta] = useField(name);
  const hasError = meta.touched && meta.error;
  const { setFieldValue } = useFormikContext();
  const areAllSelected = selectedOptions.length === options.length;

  const toggleOption = (option: Option) => {
    let updatedOptions;

    if (isSelected(option)) {
      updatedOptions = selectedOptions.filter(
        (item) => item.value !== option.value
      );
    } else {
      updatedOptions = [...selectedOptions, option];
    }
    const updatedValues = updatedOptions.map((item) => item.value);
    setSelectedOptions(updatedOptions);
    setFieldValue(name, updatedValues);
  };

  const isSelected = (option: Option) => {
    return selectedOptions.some(
      (selectedOption) => selectedOption.value === option.value
    );
  };

  useEffect(() => {
    if (areAllSelected) setIsOpen(false);
  }, [areAllSelected]);

  const renderSelectedOptions = () => (
    <ul className="flex gap-5 mt-2">
      {selectedOptions.map((option) => (
        <div className="flex items-center gap-1.5" key={option.value}>
          <Icons.dot className={circleDynamicBgStyles(option.label)} />
          <li>{option.label}</li>
          <Icons.x
            onClick={() => toggleOption(option)}
            className="cursor-pointer bg-brand-red-1000 rounded"
          />
        </div>
      ))}
    </ul>
  );

  return (
    <div className={`relative inline-block text-left w-full mb-5 ${className}`}>
      <div className="relative flex flex-col">
        <label className="ml-4 text-base font-semibold">{label}</label>
        <button
          onClick={() => setIsOpen(!isOpen)}
          type="button"
          className={`inline-flex items-center justify-between w-full px-2 h-8 py-2 text-[#323338] text-sm bg-brand-background-primary transition ${
            isOpen ? "rounded-t-md" : "rounded-md"
          }`}
          id="options-menu"
          aria-haspopup="listbox"
        >
          <p className="text-[#E0E0E1]">
            {areAllSelected ? "All of them selected" : placeholder}
          </p>
          <Icons.arrow_down
            className={`transition ${isOpen ? "rotate-180" : ""}`}
          />
        </button>
        {hasError && (
          <p className="text-brand-red-1000 text-xs ml-2 absolute -bottom-5">
            {meta.error}
          </p>
        )}
      </div>

      {isOpen && (
        <div className="absolute  w-full bg-white shadow-md z-10 rounded-b-md overflow-hidden">
          {options.map((option) => {
            // Checking if the option is selected
            const isSelected = selectedOptions.some(
              (selectedOption) => selectedOption.value === option.value
            );
            return (
              <button
                type="button"
                key={option.value}
                onClick={() => toggleOption(option)}
                className={`flex items-center justify-between px-2 py-2 text-sm text-[#323338] w-full text-left  
        } border-t first:border-none hover:text-white
        ${hoverDynamicBgStyles[option.label]}
        ${isSelected ? `text-white ${getCategoryClassName(option.label)}` : ""}
        `}
              >
                {option.label}
                {isSelected && <Icons.check />}
              </button>
            );
          })}
        </div>
      )}
      {renderSelectedOptions()}
    </div>
  );
};

export default MultiSelect;
