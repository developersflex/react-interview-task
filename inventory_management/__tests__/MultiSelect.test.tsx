import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { Form, Formik } from "formik";
import MultiSelect from "@/components/Select/MultiSelect";
import * as Yup from "yup";

describe("MultiSelect Component", () => {
  it("renders without errors", () => {
    const { container } = render(
      <Formik initialValues={{ testName: [] }} onSubmit={() => {}}>
        <MultiSelect
          options={[
            { label: "Option 1", value: "option1" },
            { label: "Option 2", value: "option2" },
          ]}
          label="Test Label"
          placeholder="Select one"
          name="testName"
        />
      </Formik>
    );

    expect(container).toMatchSnapshot();
  });

  it("toggles the dropdown on button click", () => {
    const { getByRole } = render(
      <Formik initialValues={{ testName: [] }} onSubmit={() => {}}>
        <MultiSelect
          options={[
            { label: "Option 1", value: "option1" },
            { label: "Option 2", value: "option2" },
          ]}
          label="Test Label"
          placeholder="Select one"
          name="testName"
        />
      </Formik>
    );

    const button = getByRole("button", { name: /Select one/i });
    fireEvent.click(button);

    // Assert that the dropdown menu is open
    expect(getByRole("listbox")).toBeInTheDocument();
  });

  it("closes the dropdown after all options are selected", async () => {
    const { getByRole, getByText, queryByRole } = render(
      <Formik initialValues={{ testName: [] }} onSubmit={() => {}}>
        <MultiSelect
          options={[
            { label: "Option 1", value: "option1" },
            { label: "Option 2", value: "option2" },
          ]}
          label="Test Label"
          placeholder="Select one"
          name="testName"
        />
      </Formik>
    );

    const button = getByRole("button", { name: /Select one/i });
    fireEvent.click(button);

    const option1 = getByText("Option 1");
    const option2 = getByText("Option 2");
    fireEvent.click(option1);
    fireEvent.click(option2);

    // Wait for the state to update (e.g., using waitFor)
    await waitFor(async () => {
      // Assert that the button text reflects all options being selected
      expect(button).toHaveTextContent("All of them selected");

      // Assert that the dropdown is closed
      expect(queryByRole("listbox")).not.toBeInTheDocument();
    });
  });

  it("displays error message when there are validation errors", async () => {
    const validationSchema = Yup.object().shape({
      testName: Yup.array().required("Please select an option."),
    });

    const { getByText } = render(
      <Formik
        initialValues={{ testName: "" }}
        onSubmit={() => {}}
        validationSchema={validationSchema}
      >
        <Form>
          <MultiSelect
            options={[
              { label: "Option 1", value: "option1" },
              { label: "Option 2", value: "option2" },
            ]}
            label="Test Label"
            placeholder="Select one"
            name="testName"
          />
          <button type="submit">Submit</button>
        </Form>
      </Formik>
    );

    fireEvent.click(getByText("Submit"));

    await waitFor(async () => {
      const expectedErrorMessage = "Please select an option.";
      expect(getByText(expectedErrorMessage)).toBeInTheDocument();
    });
  });
});
