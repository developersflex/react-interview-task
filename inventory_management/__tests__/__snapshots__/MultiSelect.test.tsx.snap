// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect Component renders without errors 1`] = `
<div>
  <div
    class="relative inline-block text-left w-full mb-5 undefined"
  >
    <div
      class="relative flex flex-col"
    >
      <label
        class="ml-4 text-base font-semibold"
      >
        Test Label
      </label>
      <button
        aria-haspopup="listbox"
        class="inline-flex items-center justify-between w-full px-2 h-8 py-2 text-[#323338] text-sm bg-brand-background-primary transition rounded-md"
        id="options-menu"
        type="button"
      >
        <p
          class="text-[#E0E0E1]"
        >
          Select one
        </p>
        <svg
          class="transition "
          fill="none"
          height="8"
          viewBox="0 0 12 8"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.27378 2.85924L5.31178 6.89324C5.45605 7.03801 5.62749 7.15287 5.81625 7.23125C6.00501 7.30962 6.20739 7.34997 6.41178 7.34997C6.61616 7.34997 6.81854 7.30962 7.0073 7.23125C7.19606 7.15287 7.3675 7.03801 7.51178 6.89324L11.5498 2.85524C11.766 2.63592 11.9126 2.35765 11.9713 2.05532C12.03 1.75299 11.9982 1.44006 11.8797 1.15577C11.7613 0.871477 11.5616 0.628481 11.3056 0.45725C11.0496 0.286019 10.7487 0.194176 10.4408 0.193237L2.36478 0.193237C2.05645 0.194567 1.75544 0.287293 1.49981 0.459688C1.24419 0.632084 1.04542 0.876408 0.928641 1.16177C0.811865 1.44712 0.782327 1.7607 0.843759 2.06285C0.905192 2.36499 1.05484 2.64214 1.27378 2.85924Z"
            fill="#323338"
          />
        </svg>
      </button>
    </div>
    <ul
      class="flex gap-5 mt-2"
    />
  </div>
</div>
`;
